Sources/Core/Futures/Future.swift

```diff
@@ -6,241 +6,71 @@ import Dispatch
-    }
-    
-    /// Creates a new, uncompleted, unprovoked future
-    internal init() {}
-    
-    /// Creates a future, executing asynchronously on either the provided or global queue
-    ///
-    /// - parameter queue: If provided, will execute the closure on this queue
-    /// - parameter closure: The closure to execute.
-    public init(queue: DispatchQueue? = nil, _ closure: @escaping (() throws -> (Expectation))) {
-        let queue = queue ?? DispatchQueue.global()
-        
-        queue.async {
-            do {
-                self.expectation = try closure()
-            } catch {
-                self.error = error
-            }
-            
-            self.complete()
-        }
-    }
-    
-    /// Creates a new future by transforming one future into a new future.
-    ///
-    /// The post-transform future will become this future
-    internal init<Base>(transform: @escaping ((Base) throws -> (Future<Expectation>)), from: Future<Base>) {
-        from.onComplete(asynchronously: nil) { result in
-            switch result {
-            case .expectation(let data):
-                do {
-                    let promise = try transform(data)
-                    
-                    promise.onComplete(asynchronously: nil) { result in
-                        switch result {
-                        case .expectation(let expectation):
-                            self.expectation = expectation
-                        case .error(let error):
-                            self.error = error
-                        }
-                        
-                        self.complete()
-                    }
-                } catch {
-                    self.error = error
-                    self.complete()
-                }
-            case .error(let error):
-                self.error = error
-                self.complete()
-            }
-        }
-    }
-    
-    /// Creates a new future by transforming one future's results into another result.
-    ///
-    /// The post-transform result will be this future's result.
-    internal init<Base>(transform: @escaping ((Base) throws -> (Expectation)), from: Future<Base>) {
-        from.onComplete(asynchronously: nil) { result in
-            switch result {
-            case .expectation(let data):
-                do {
-                    self.expectation = try transform(data)
-                } catch {
-                    self.error = error
-                }
-            case .error(let error):
-                self.error = error
+
+        if let result = self.result {
+            queue.async {
+                callback(result)
             }
-            
-            self.complete()
+        } else {
+            let awaiter = Awaiter(callback: callback, queue: queue)
+            awaiters.append(awaiter)
         }
     }
-    
-    public func map<B>(_ closure: @escaping ((Expectation) throws -> (B))) -> Future<B> {
```

Comment:
This was a really important part of keeping the API clean with regards to managing errors in combination with `then` handlers, too.